observatory-dags/observatory/dags/telescopes/jstor.py

```diff
@@ -0,0 +1,396 @@
+import datetime
+import logging
+import os
+import os.path
+import os.path
+import time
+from collections import OrderedDict
+from datetime import datetime
+from typing import Dict
+from typing import List
+
+import pendulum
+import requests
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+from airflow.models.taskinstance import TaskInstance
+from bs4 import BeautifulSoup, SoupStrainer
+from google.cloud.bigquery import SourceFormat
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import Resource, build
+from observatory.platform.telescopes.snapshot_telescope import SnapshotRelease, SnapshotTelescope
+from observatory.platform.utils.airflow_utils import AirflowConns, AirflowVars
+from observatory.platform.utils.telescope_utils import convert
+from observatory.platform.utils.telescope_utils import list_to_jsonl_gz
+from observatory.platform.utils.template_utils import upload_files_from_list
+from pendulum import Pendulum
+
+
+class JstorRelease(SnapshotRelease):
+    def __init__(self, dag_id: str, release_date: Pendulum, reports_info: List[dict]):
+        """ Construct a JstorRelease.
+
+        :param release_date: the release date.
+        :param reports_info: list with report_type (country or institution) and url of reports
+        """
+
+        self.reports_info = reports_info
+        download_files_regex = f"^{dag_id}_(country|institution)\.tsv$"
+        transform_files_regex = f"^{dag_id}_(country|institution)\.jsonl.gz"
+
+        super().__init__(dag_id, release_date, download_files_regex, transform_files_regex)
+
+    def download_path(self, report_type: str) -> str:
+        """ Creates full download path
+
+        :param report_type: The report type (country or institution)
+        :return: Download path
+        """
+        return os.path.join(self.download_folder, f'{self.dag_id}_{report_type}.tsv')
+
+    def transform_path(self, report_type: str) -> str:
+        """ Creates full transform path
+
+        :param report_type: The report type (country or institution)
+        :return: Transform path
+        """
+        return os.path.join(self.transform_folder, f"{self.dag_id}_{report_type}.jsonl.gz")
+
+    def download(self):
+        """ Downloads Google Books reports.
+
+        :return: the paths on the system of the downloaded files.
+        """
+        for report in self.reports_info:
+            headers = create_headers(report['url'])
+            time.sleep(20)
+            response = requests.get(report['url'], headers=headers)
+            if response.status_code != 200:
+                raise AirflowException(f'Could not get content from download url, reason: {response.reason}, '
+                                       f'status_code: {response.status_code}')
+
+            content = response.content.decode('utf-8')
+            with open(self.download_path(report['type']), 'w') as f:
+                f.write(content)
```

Comment:
Instead of just waiting for 20 seconds for each download and failing the whole task if one them doesn't download, it would be good to add functionality to retry the download. You can use a library called tenacity, which is already installed with Airflow. We can only use version 4.12.0 however. It would also be good to put variables such as the fixed sleep time into a constant variable. You would need to put the code for downloading the report into it's own function and example is below:

```
from datetime import datetime

from airflow import AirflowException
from tenacity import retry, stop_after_attempt, wait_exponential, wait_fixed

MAX_ATTEMPTS = 3
FIXED_WAIT = 20  # seconds
MAX_WAIT_TIME = 60 * 10  # seconds
EXP_BASE = 3
MULTIPLIER = 10


@retry(
    wait=wait_fixed(FIXED_WAIT)
    + wait_exponential(multiplier=MULTIPLIER, exp_base=EXP_BASE, max=MAX_WAIT_TIME),
    stop=stop_after_attempt(MAX_ATTEMPTS),
    reraise=True,
)
def download_report(status_code):
    print(f"Try {datetime.now().strftime('%H:%M:%S')}")

    if status_code != 200:
        raise AirflowException("API response: {}".format(status_code))
    return "Done!"


for i in range(32):
    download_report(201)
```

Reply:
Thanks that was very useful, I've added this. I left the values for the variables the same as in the example, they seem reasonable to me.